src/item.h

```diff
@@ -206,6 +207,226 @@ class ItemAttributes
 			return static_cast<ItemDecayState_t>(getIntAttr(ITEM_ATTRIBUTE_DECAYSTATE));
 		}
 
+		struct CustomAttributeKey
+		{
+			union {
+				std::string* string;
+				int64_t integer;
+			} value;
+
+			customAttrTypes type = ATTR_NO_TYPE;
+
+			explicit CustomAttributeKey(customAttrTypes type) : type(type) {
+				memset(&value, 0, sizeof(value));
+			}
+
+			CustomAttributeKey(const std::string& v) {
+				type = ATTR_STRING_TYPE;
+				value.string = new std::string(v);
+			}
+
+			CustomAttributeKey(const char* v) {
+				type = ATTR_STRING_TYPE;
+				value.string = new std::string(v);
+			}
+
+			CustomAttributeKey(int64_t v) {
+				type = ATTR_INTEGER_TYPE;
+				value.integer = v;
+			}
+
+			CustomAttributeKey(const CustomAttributeKey& i) {
+				type = i.type;
+				switch (type) {
+					case ATTR_STRING_TYPE: {
+						value.string = new std::string(*i.value.string);
+					}
+
+					case ATTR_INTEGER_TYPE: {
+						value.integer = i.value.integer;
+					}
+
+					default: {
+						memset(&value, 0, sizeof(value));
+					}
+				}
+			}
+
+			CustomAttributeKey(CustomAttributeKey&& attribute) : value(attribute.value), type(attribute.type) {
+				memset(&attribute.value, 0, sizeof(value));
+				attribute.type = ATTR_NO_TYPE;
+			}
+
+			~CustomAttributeKey() {
+				if (type == ATTR_STRING_TYPE) {
+					delete value.string;
+				}
+			}
+
+			CustomAttributeKey& operator=(CustomAttributeKey other) {
+				CustomAttributeKey::swap(*this, other);
+				return *this;
+			}
+
+			CustomAttributeKey& operator=(CustomAttributeKey&& other) {
+				if (this != &other) {
+					if (type == ATTR_STRING_TYPE) {
+						delete value.string;
+					}
+
+					value = other.value;
+					type = other.type;
+					memset(&other.value, 0, sizeof(value));
+					other.type = ATTR_NO_TYPE;
+				}
+				return *this;
+			}
+
+			bool operator==(const CustomAttributeKey& other) const {
+				if (type != other.type) {
+					return false;
+				}
+				else {
+					switch (type) {
+						case ATTR_INTEGER_TYPE: {
+							return value.integer == other.value.integer;
+						}
+
+						case ATTR_STRING_TYPE: {
+							return !value.string->compare(*other.value.string);
+						}
+
+						default: {
+							return false;
+						}
+					}
+				}
+			}
+
+			static void swap(CustomAttributeKey& first, CustomAttributeKey& second) {
+				std::swap(first.value, second.value);
+				std::swap(first.type, second.type);
+			}
+
+			struct hash {
+				std::size_t operator()(const CustomAttributeKey& key) const {
+					switch (key.type) {
+						case ATTR_INTEGER_TYPE: {
+							return std::hash<int64_t>()(key.value.integer);
+						}
+
+						case ATTR_STRING_TYPE: {
+							return std::hash<std::string>()(*key.value.string);
+						}
+
+						default: {
+							return 0;
+						}
+					}
+				}
+			};
+		};
+
+		struct CustomAttribute
+		{
+			union {
+				std::string* string;
+				int64_t integer;
+				double doublen;
+				bool boolean;
+			} value;
```

Comment:
This union should be a separate class (or preferably use boost:any/boost::variant instead)